contracts/command-batcher/src/contract.rs

```diff
@@ -1,44 +1,343 @@
+            execute::rotate_snapshot(deps, env, info, pub_keys)
+        }
     }
 }
 
 pub mod execute {
     use super::*;
 
-    pub fn construct_proof(_message_ids: Vec<String>) -> Result<Response, ContractError> {
-        todo!()
+    pub fn construct_proof(
+        deps: DepsMut,
+        message_ids: Vec<String>,
+    ) -> Result<Response, ContractError> {
+        let config = CONFIG.load(deps.storage)?;
+
+        let query = gateway::msg::QueryMsg::GetMessages { message_ids };
+        let messages: Vec<Message> = deps.querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
+            contract_addr: config.gateway.into(),
+            msg: to_binary(&query)?,
+        }))?;
+
+        if messages.is_empty() {
+            return Err(ContractError::NoMessagesFound {});
+        }
+
+        let message_ids: Vec<String> = messages.iter().map(|msg| msg.id.clone()).collect();
+        let batch_id = BatchID::new(&message_ids);
+
+        let command_batch = match COMMANDS_BATCH.may_load(deps.storage, &batch_id)? {
+            Some(batch) => batch,
+            None => {
+                let batch = CommandBatch::new(messages, config.destination_chain_id)?;
+
+                COMMANDS_BATCH.save(deps.storage, &batch.id, &batch)?;
+
+                batch
+            }
+        };
+
+        let reply_id = REPLY_ID_COUNTER.update(deps.storage, |mut reply_id| -> StdResult<_> {
+            reply_id += 1;
+            Ok(reply_id)
+        })?;
+        REPLY_ID_TO_BATCH.save(deps.storage, reply_id, &command_batch.id)?;
+
+        let start_sig_msg = multisig::msg::ExecuteMsg::StartSigningSession {
+            key_id: config.service_name,
+            msg: command_batch.msg_to_sign(),
+        };
+
+        let wasm_msg = wasm_execute(config.multisig, &start_sig_msg, vec![])?;
+
+        Ok(Response::new().add_submessage(SubMsg::reply_on_success(wasm_msg, reply_id)))
+    }
+
+    pub fn rotate_snapshot(
+        deps: DepsMut,
+        env: Env,
+        info: MessageInfo,
+        pub_keys: HashMap<String, HexBinary>,
+    ) -> Result<Response, ContractError> {
+        let config = CONFIG.load(deps.storage)?;
+        if config.admin != info.sender {
+            return Err(ContractError::Unauthorized {});
+        }
+
+        let snapshot = snapshot(deps.querier, env.block, &config)?;
+
+        for participant in snapshot.participants.keys() {
+            if !pub_keys.contains_key(participant) {
+                return Err(ContractError::PublicKeyNotFound {
+                    participant: participant.to_owned(),
+                });
+            }
+        }
+
+        let keygen_msg = multisig::msg::ExecuteMsg::KeyGen {
+            key_id: config.service_name,
+            snapshot,
+            pub_keys,
+        };
+
+        Ok(Response::new().add_message(WasmMsg::Execute {
+            contract_addr: config.multisig.into(),
+            msg: to_binary(&keygen_msg)?,
+            funds: vec![],
+        }))
```

Comment:
Need an event here

Reply:
I think we should add still add some event here, especially because these contracts are all changing quite a bit. I don't think we need the owner, but the other attributes are good.